actors/builtin/market/deal.go

```diff
@@ -12,6 +24,167 @@ import (
 //}
 var PieceCIDPrefix = market0.PieceCIDPrefix
 
+// The DealLabel is a kinded union of string or byte slice.
+// It serializes to a CBOR string or CBOR byte string depending on which form it takes.
+// The empty label is serialized as an empty CBOR string (maj type 3).
+type DealLabel struct {
+	bs        []byte
+	notString bool
+}
+
+// Zero value of DealLabel is canonical EmptyDealLabel but three different values serialize the same way:
+// DealLabel{nil, nil}, (*DealLabel)(nil), and DealLabel{&"", nil}
+var EmptyDealLabel = DealLabel{}
+
+func NewLabelFromString(s string) (DealLabel, error) {
+	if len(s) > DealMaxLabelSize {
+		return EmptyDealLabel, xerrors.Errorf("provided string is too large to be a label (%d), max length (%d)", len(s), DealMaxLabelSize)
+	}
+	if !utf8.ValidString(s) {
+		return EmptyDealLabel, xerrors.Errorf("provided string is invalid utf8")
+	}
+	return DealLabel{
+		bs:        []byte(s),
+		notString: false,
+	}, nil
+}
+
+func NewLabelFromBytes(b []byte) (DealLabel, error) {
+	if len(b) > DealMaxLabelSize {
+		return EmptyDealLabel, xerrors.Errorf("provided bytes are too large to be a label (%d), max length (%d)", len(b), DealMaxLabelSize)
+	}
+
+	return DealLabel{
+		bs:        b,
+		notString: true,
+	}, nil
+}
+
+func (label DealLabel) IsString() bool {
+	return !label.notString
+}
+
+func (label DealLabel) IsBytes() bool {
+	return label.notString
+}
+
+func (label DealLabel) ToString() (string, error) {
+	if !label.IsString() {
+		return "", xerrors.Errorf("label is not string")
+	}
+
+	return string(label.bs), nil
+}
+
+func (label DealLabel) ToBytes() ([]byte, error) {
+	if !label.IsBytes() {
+		return nil, xerrors.Errorf("label is not bytes")
+	}
+	return label.bs, nil
+}
+
+func (label DealLabel) Length() int {
+	return len(label.bs)
+}
+
+func (l DealLabel) Equals(o DealLabel) bool {
+	return bytes.Equal(l.bs, o.bs) && l.notString == o.notString
+}
+
+func (label *DealLabel) MarshalCBOR(w io.Writer) error {
```

Comment:
I think we can vastly simplify this now:

1. If null, write empty string.
2. If too large, return an error.
3. Set a `majorType` variable depending on the type.
4. Write the value (no need for a conditional).